Pass bilingual_str argument to AbortNode()

hebasto opened the pull request:
This PR is a [followup](https://github.com/bitcoin/bitcoin/issues/16218#issuecomment-625919724) of #16224, and it adds `bilingual_str` type argument support to the `AbortNode()` functions.

hebasto committed:
refactor: Use bilingual_str::empty()

hebasto committed:
Pass bilingual_str argument to AbortNode()

hebasto committed:
Drop MSG_NOPREFIX flag

Since bilingual_str type is fully supported, the MSG_NOPREFIX flag is no
longer needed.

hebasto commented:
This PR could be tested with a patch similar to the following:
```diff
diff --git a/src/qt/locale/bitcoin_nl.ts b/src/qt/locale/bitcoin_nl.ts
index 73c18cd4d..13bda2db2 100644
--- a/src/qt/locale/bitcoin_nl.ts
+++ b/src/qt/locale/bitcoin_nl.ts
@@ -3524,8 +3524,8 @@ Notitie: Omdat de vergoeding per byte wordt gerekend, zal een vergoeding van "10
         <translation>P2P-adressen aan het laden...</translation>
     </message>
     <message>
-        <source>Error: Disk space is too low!</source>
-        <translation>Error: Opslagruimte te weinig!</translation>
+        <source>Disk space is too low!</source>
+        <translation>Opslagruimte te weinig!</translation>
     </message>
     <message>
         <source>Loading banlist...</source>
diff --git a/src/validation.cpp b/src/validation.cpp
index 9034f66e9..c9539c325 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -2317,7 +2317,7 @@ bool CChainState::FlushStateToDisk(
         // Write blocks and block index to disk.
         if (fDoFullFlush || fPeriodicWrite) {
             // Depend on nMinDiskSpace to ensure we can write block index
-            if (!CheckDiskSpace(GetBlocksDir())) {
+            if (CheckDiskSpace(GetBlocksDir())) {
                 return AbortNode(state, "Disk space is too low!", _("Disk space is too low!"));
             }
             {
```

```
$ ./src/qt/bitcoin-qt -testnet -lang=nl
Error: Disk space is too low!
Error: Disk space is too low!
Error: Disk space is too low!
```

![Screenshot from 2020-05-10 12-55-36](https://user-images.githubusercontent.com/32963518/81496390-1ad89880-92c0-11ea-9230-10b9e33abb87.png)

```
$ tail -4 ~/.bitcoin/testnet3/debug.log 
2020-05-10T09:55:41Z [shutoff] *** Disk space is too low!
2020-05-10T09:55:41Z [shutoff] Error: Disk space is too low!
2020-05-10T09:55:41Z [shutoff] ForceFlushStateToDisk: failed to flush state (Disk space is too low!)
2020-05-10T09:55:41Z [shutoff] Shutdown: done
```

MarcoFalke commented on src/util/translation.h:
```diff
@@ -16,6 +16,11 @@
+
+    bool empty() const
+    {
+        return original.empty() && translated.empty();
```
```suggestion
        return original.empty();
```

I think translators should not be able to influence executions paths by blanking messages

hebasto replied:
> I think translators should not be able to influence executions paths by blanking messages

Sure!

hebasto replied:
[Update](https://github.com/bitcoin/bitcoin/pull/18927#issuecomment-626321249).

hebasto commented:
Updated 1511dd0b1a44727be652b61256899b0ce8024178 -> 6fad0cf7a9b9492e92f6331db4cd6fcd14977d5e ([pr18927.02](https://github.com/hebasto/bitcoin/commits/pr18927.02) -> [pr18927.03](https://github.com/hebasto/bitcoin/commits/pr18927.03), [diff](https://github.com/hebasto/bitcoin/compare/pr18927.02..pr18927.03)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/18927#discussion_r422638039):
> I think translators should not be able to influence executions paths by blanking messages

hebasto commented:
Updated 6fad0cf7a9b9492e92f6331db4cd6fcd14977d5e -> 4067da94bbf24c6b2b9d45808ae33dc9e7b9d6d4 ([pr18927.03](https://github.com/hebasto/bitcoin/commits/pr18927.03) -> [pr18927.04](https://github.com/hebasto/bitcoin/commits/pr18927.04), [diff](https://github.com/hebasto/bitcoin/compare/pr18927.03..pr18927.04)):

- more uses of `bilingual_str::empty()`

hebasto commented:
Rebased 4067da94bbf24c6b2b9d45808ae33dc9e7b9d6d4 -> c5e34e58801654028e3039188e2882e25df594a4 ([pr18927.04](https://github.com/hebasto/bitcoin/commits/pr18927.04) -> [pr18927.05](https://github.com/hebasto/bitcoin/commits/pr18927.05)) due to the conflict with #18922.

MarcoFalke commented:
Needs rebase

hebasto commented:
@MarcoFalke 
> Needs rebase

Rebased c5e34e58801654028e3039188e2882e25df594a4 -> d924f2a596c8f37deb2dd94069c578244823c31f ([pr18927.05](https://github.com/hebasto/bitcoin/commits/pr18927.05) -> [pr18927.07](https://github.com/hebasto/bitcoin/commits/pr18927.07)).

MarcoFalke commented on src/validation.cpp:
```diff
@@ -1655,23 +1655,22 @@ bool UndoReadFromDisk(CBlockUndo& blockundo, const CBlockIndex* pindex)
+        uiInterface.ThreadSafeMessageBox(userMessage, "", CClientUIInterface::MSG_ERROR);
     } else {
-        uiInterface.ThreadSafeMessageBox(_("Error: A fatal internal error occurred, see debug.log for details"), "", CClientUIInterface::MSG_ERROR | CClientUIInterface::MSG_NOPREFIX);
+        uiInterface.ThreadSafeMessageBox(_("A fatal internal error occurred, see debug.log for details"), "", CClientUIInterface::MSG_ERROR);
```
Now that the signature is identical to InitError, the call could be simplified to just a call to `AbortError`, which would be a new alias for InitError?

```cpp
constexpr auto AbortError = InitError;

hebasto replied:
> Now that the signature is identical to InitError...

Which function signature do you mean?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18927#issuecomment-640708390).

MarcoFalke commented on src/validation.cpp:
```diff
@@ -1655,23 +1655,22 @@ bool UndoReadFromDisk(CBlockUndo& blockundo, const CBlockIndex* pindex)
+        uiInterface.ThreadSafeMessageBox(userMessage, "", CClientUIInterface::MSG_ERROR);
     } else {
-        uiInterface.ThreadSafeMessageBox(_("Error: A fatal internal error occurred, see debug.log for details"), "", CClientUIInterface::MSG_ERROR | CClientUIInterface::MSG_NOPREFIX);
+        uiInterface.ThreadSafeMessageBox(_("A fatal internal error occurred, see debug.log for details"), "", CClientUIInterface::MSG_ERROR);
```
All that InitError does is pass a translated string into `ThreadSafeMessageBox`, so this line could be replaced by a simple call to InitError, no?

```suggestion
        AbortError(_("A fatal internal error occurred, see debug.log for details"));
```

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/18927#issuecomment-640708390).

hebasto committed:
refactor: Add AbortError alias

hebasto commented:
Updated d924f2a596c8f37deb2dd94069c578244823c31f -> 5527be06277647dffe7cda587c4bbfbec2a5c8ca ([pr18927.07](https://github.com/hebasto/bitcoin/commits/pr18927.07) -> [pr18927.08](https://github.com/hebasto/bitcoin/commits/pr18927.08), [diff](https://github.com/hebasto/bitcoin/compare/pr18927.07..pr18927.08)):

- implemented @MarcoFalke's suggestion (https://github.com/bitcoin/bitcoin/pull/18927#discussion_r436357871, https://github.com/bitcoin/bitcoin/pull/18927#discussion_r436608141)

MarcoFalke commented:
ACK 5527be06277647dffe7cda587c4bbfbec2a5c8ca 👟

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 5527be06277647dffe7cda587c4bbfbec2a5c8ca 👟
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUi2JQwAwwhgq56EfM6hpOzafEWzJZZinddkJuSLzeiDHy1b7SUgQVooQWpqBqiJ
KaI87+DzkSGBOu0vqoy2DjuuFcyLpHSdOQ8fAJFOYbfxbx6M37bY7Hch+eeYivjt
k5gPNiVBPzz6h68fxgPNshwHVR6qWzI5RgOe8/GsYrPbGjh42ueYgVZcHdxgqHEP
vjHJzkU31LW3Ntqg1P6VuYt8uxGvXiejQOjACs9FxWgtemZMRnvHSrUIGfWNyRPt
G9Y2ClnpZ5rhy6CW3FlyCPEtY6GofbQ9AUEUo/R6Dxv4xVk9xBw1wJ0vwV7q/clW
OjTQgr8LK8puSiGHAhc3ee+3n+hsDmIuzfY4xLv3E4Bqe7/jbvlyNzQe/UfAiGSK
8PqZp0jk+ZG87qZuwNKYytvXDPNOAorN1SO8bIWUlt0cTlfKfQcdHnHAQYMdUxkl
Xsu5LcmEgyuerJ7pPQCQ5gVYpCScCNjzwvOIKdkggY1zBJNB2YkdP3aKMYU+ezdL
UbthjJ2Q
=zsXO
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `c4a6904cff17d24ba7d628d970246ed0ce8d86a0caa4d68859d702cd905c7c6a  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108c4a6904cff17d24ba7d628d970246ed0ce8d86a0caa4d68859d702cd905c7c6af010af587135f7d85302f3f2e256cc2e718b08fff01039950d2c5e7832ed2a9f53f751c7391f08f1045ee00453f008a03a2c8296de62bb0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01064d880d477d64cca5f01b8bf7b1d58d608f1045ee00453f0084405133862e512b70083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0106da3e355353421acf9dbf86bf171df4a08f1045ee00453f00881173fa7fa3f26ea0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f48a814049783dd76548ed4424d2895608f1045ee00453f008f4f6b1f24d06b74e0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>
